### DISCUSSION TITLE
Where does Plenti sit in its rendering patterns for web apps?

### DISCUSSION THREAD

**Discussion by jamestagal** [plentico/plenti#257] [Q&A] [answered]
Hi all,

I ask this question as way to better understand this technology.  I was watching this great video from Jeff Delaney on [10 Rendering Patterns for Web Apps](https://www.youtube.com/watch?v=Dkx5ydvtpCA) this morning and thinking about Plenti. Which rendering patterns it uses...I think it should be his number 5. Static Site Generation with some Hydration ? 

I see that Plenti is now a [Build-Time Render (BTR) Engine](https://plenti.co/docs) and also may be referred to as (“static”) rendering because it is precomputed server-side rendering. This article [Theoretical foundations for server-side rendering and static-rendering](https://htmlpreview.github.io/?https://github.com/lbke/ssr-theory/blob/main/readme.html)  although heavy in parts was helpful for me to further understand the different rendering patterns.

Also interestingly I see that [SvelteKit 1.0](https://kit.svelte.dev/) was released recently which includes a lot of these rendering patterns. I think SvelteKit is like a backend framework that uses Svelte as its component framework. Below are the main Svetltekit features I could see:

- File based routing 
- Configurable Rendering (SSG, SSR, CSR)
- Data fetching
- Code splitting (loading data on demand) - Is this pre-loading?
- Deployment options - using adapters? 
- Progressive enhancement - Is this to incrementally enable javascript where required ?

Let me know what you think about any of the above.

Cheers,
Ben

**Comment by jimafisk** [comment, marked as answer]
We have aspects of several rendering strategies. Fundamentally Plenti creates [static websites](https://youtu.be/Dkx5ydvtpCA?t=39), but that's the fallback behavior because Plenti also automatically hydrates into an [SPA](https://youtu.be/Dkx5ydvtpCA?t=112). That's exactly the technique Jeff describes as [server-side rendering with hydration](https://youtu.be/Dkx5ydvtpCA?t=166), but the real differentiation is _when_ this process happens. He's discussing the SSR taking place on each web request, in Plenti the SSR takes place at build time... and as I watch the video further I see he mentions [SSG with hydration](https://youtu.be/Dkx5ydvtpCA?t=204) - this is exactly what we're doing. So you're 100% right @jamestagal, it's number 5.

> I see that Plenti is now a Build-Time Render (BTR) Engine

That's just me trying to popularize a different way to describe these type of apps because I wasn't happy with the current terminology (because of course we need more acronyms :joy:). Other existing terms for this are SSG and Jamstack, but IMO those do a poor job describing what differentiates these apps:
- SSG implies the site is "static" when in reality it can be quite dynamic. Not only with interactive frontends, but with APIs and even content editing like we're doing with a git-backed CMS.
- JAM refers to JavaScript, APIs, and Markup - almost every website uses these concepts, but not all those sites would tecnically be Jamstack sites. Where they do their rendering is what actually makes them Jamstack.

>  see that SvelteKit 1.0 was released recently which includes a lot of these rendering patterns

I haven't had a ton of time to play around with SvelteKit, but it's safe to bet it does a lot more than Plenti does. With SvelteKit you can choose among several different rending strategies vs with Plenti we are targeting one very specific way to build websites because we think it's the easiest and cheapest way possible to run these sort of apps. We basically want to bring super powers to frontend developers so they can design an interface and essentially get the backend / editing experience for free without having to rig it up themselves.

**Reply by jamestagal** [reply]
Thanks @jimafisk 

>That's just me trying to popularize a different way to describe these type of apps because I wasn't happy with the current terminology

I like it!  **Build-Time Render (BTR) Engine**  does seem to capture more of the essence of what it is actually doing. I will help you popularise it 😸 
>We basically want to bring super powers to frontend developers so they can design an interface and essentially get the backend / editing experience for free without having to rig it up themselves.

Yes I see the "super powers" you are enabling for developers.  I will try to do more marketing on Twitter when I can and think about of other platforms also. Also, I would be happy to get on a call with you (for your Youtube channel) after we finish building the ```educenter``` theme and talk through some of the features we are setting up, tips for building and lessons learned etc.. 

Ben

**Reply by jimafisk** [reply]
That would be great @jamestagal! I was thinking it might be cool to hold a monthly Plenti call that anyone could join to discuss:
- Projects they're working on
- Questions they have
- Bugs they've found
- Roadmap / vision

We could record these and post them to youtube for the async folks. I'm all out on a project at the moment, but I'd like to target starting these in May.
